roachtest: clearrange/checks=true/rangeTs=false failed

roachtest.clearrange/checks=true/rangeTs=false [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10689032?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10689032?buildTab=artifacts#/clearrange/checks=true/rangeTs=false) on master @ [0b0a212ecbe514b55301f4a963edc92b16de9bab](https://github.com/cockroachdb/cockroach/commits/0b0a212ecbe514b55301f4a963edc92b16de9bab):


```
(cluster.go:2279).Run: output in run_120634.914037609_n1-10_cockroach-workload-r: ./cockroach workload run kv --concurrency=32 --duration=1h returned: COMMAND_PROBLEM: exit status 1
(monitor.go:137).Wait: monitor failure: monitor task failed: t.Fatal() was called
test artifacts and logs in: /artifacts/clearrange/checks=true/rangeTs=false/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=16</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_fs=ext4</code>
, <code>ROACHTEST_localSSD=true</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*clearrange/checks=true/rangeTs=false.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29127

```
panic: runtime error: index out of range [-128]

goroutine 789 [running]:
panic({0x5d60140, 0xc009913080})
	GOROOT/src/runtime/panic.go:987 +0x3ba fp=0xc03a792010 sp=0xc03a791f50 pc=0x49e43a
runtime.goPanicIndex(0xffffffffffffff80, 0x5)
	GOROOT/src/runtime/panic.go:113 +0x7f fp=0xc03a792050 sp=0xc03a792010 pc=0x49c25f
github.com/cockroachdb/cockroach/pkg/util/admission.(*GrantCoordinator).tryGrantLocked(0xc00702e840)
	github.com/cockroachdb/cockroach/pkg/util/admission/grant_coordinator.go:896 +0x2c5 fp=0xc03a7920d8 sp=0xc03a792050 pc=0x1699b25
github.com/cockroachdb/cockroach/pkg/util/admission.(*kvStoreTokenGranter).storeReplicatedWorkAdmittedLocked(0xc0016600c0, 0x0, 0x82, {0x0?, 0x0?})
	github.com/cockroachdb/cockroach/pkg/util/admission/granter.go:582 +0x13f fp=0xc03a792120 sp=0xc03a7920d8 pc=0x169d35f
github.com/cockroachdb/cockroach/pkg/util/admission.(*kvStoreTokenChildGranter).storeReplicatedWorkAdmittedLocked(0x97e6ab?, 0x0?, {0x0?, 0x0?})
	github.com/cockroachdb/cockroach/pkg/util/admission/granter.go:374 +0x30 fp=0xc03a792158 sp=0xc03a792120 pc=0x169c650
github.com/cockroachdb/cockroach/pkg/util/admission.(*StoreWorkQueue).admittedReplicatedWork(0xc00707c000, {0x14dddd96f04?}, 0x0, {0x1, 0x9ab, 0x5, {0x6, 0x19ce2e}, 0x0}, 0x82, ...)
	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:1990 +0x1b8 fp=0xc03a7921e0 sp=0xc03a792158 pc=0x16aa3f8
github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).granted(0xc00707c008, 0x0)
```

Same as #105185.

--- a/pkg/kv/kvserver/flow_control_integration_test.go
+++ b/pkg/kv/kvserver/flow_control_integration_test.go
@@ -2203,6 +2203,117 @@ ORDER BY name ASC;
 `)
 }
 
+// TestFlowControlGranterAdmitOneByOne is a reproduction for #105185. Internal
+// admission code that relied on admitting at most one waiting request was in
+// fact admitting more than one, and doing so recursively with call stacks as
+// deep as the admit chain. This triggered panics (and is also just undesirable,
+// design-wise). This test intentionally queues a 1000+ small requests, to that
+// end.
+func TestFlowControlGranterAdmitOneByOne(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	const numNodes = 3
+	var disableWorkQueueGranting atomic.Bool
+	disableWorkQueueGranting.Store(true)
+
+	st := cluster.MakeTestingClusterSettings()
+	kvflowcontrol.Enabled.Override(ctx, &st.SV, true)
+
+	tc := testcluster.StartTestCluster(t, numNodes, base.TestClusterArgs{
+		ReplicationMode: base.ReplicationManual,
+		ServerArgs: base.TestServerArgs{
+			Settings: st,
+			Knobs: base.TestingKnobs{
+				Store: &kvserver.StoreTestingKnobs{
+					FlowControlTestingKnobs: &kvflowcontrol.TestingKnobs{
+						UseOnlyForScratchRanges: true,
+						MaintainStreamsForBehindFollowers: func() bool {
+							// TODO(irfansharif): This test is flakey without
+							// this change -- we disconnect one stream or
+							// another because raft says we're no longer
+							// actively replicating through it. Why? Something
+							// to do with the many proposals we're issuing?
+							return true
+						},
+						OverrideTokenDeduction: func() kvflowcontrol.Tokens {
+							// This test asserts on the exact values of tracked
+							// tokens. In non-test code, the tokens deducted are
+							// a few bytes off (give or take) from the size of
+							// the proposals. We don't care about such
+							// differences.
+							return kvflowcontrol.Tokens(1 << 10 /* 1KiB */)
+						},
+					},
+				},
+				AdmissionControl: &admission.TestingKnobs{
+					DisableWorkQueueFastPath: true,
+					DisableWorkQueueGranting: func() bool {
+						return disableWorkQueueGranting.Load()
+					},
+					AlwaysTryGrantWhenAdmitted: true,
+				},
+			},
+		},
+	})
+	defer tc.Stopper().Stop(ctx)
+
+	k := tc.ScratchRange(t)
+	tc.AddVotersOrFatal(t, k, tc.Targets(1, 2)...)
+
+	n1 := sqlutils.MakeSQLRunner(tc.ServerConn(0))
+
+	h := newFlowControlTestHelper(t, tc)
+	h.init()
+	defer h.close("granter_admit_one_by_one")
+
+	desc, err := tc.LookupRange(k)
+	require.NoError(t, err)
+	h.waitForConnectedStreams(ctx, desc.RangeID, 3)
+
+	h.comment(`-- (Issuing regular 1024*1KiB, 3x replicated writes that are not admitted.)`)
+	h.log("sending put requests")
+	for i := 0; i < 1024; i++ {
+		h.put(ctx, k, 1<<10 /* 1KiB */, admissionpb.NormalPri)
+	}
+	h.log("sent put requests")
+
+	h.comment(`
+-- Flow token metrics from n1 after issuing 1024KiB, i.e. 1MiB 3x replicated writes
+-- that are yet to get admitted. We see 3*1MiB=3MiB deductions of
+-- {regular,elastic} tokens with no corresponding returns.
+`)
+	h.query(n1, `
+  SELECT name, crdb_internal.humanize_bytes(value::INT8)
+    FROM crdb_internal.node_metrics
+   WHERE name LIKE '%kvadmission%tokens%'
+ORDER BY name ASC;
+`)
+
+	h.comment(`-- Observe the total tracked tokens per-stream on n1.`)
+	h.query(n1, `
+  SELECT range_id, store_id, crdb_internal.humanize_bytes(total_tracked_tokens::INT8)
+    FROM crdb_internal.kv_flow_control_handles
+`, "range_id", "store_id", "total_tracked_tokens")
+
+	h.comment(`-- (Allow below-raft admission to proceed.)`)
+	disableWorkQueueGranting.Store(false)
+	h.waitForAllTokensReturned(ctx, 3) // wait for admission
+
+	h.comment(`
+-- Flow token metrics from n1 after work gets admitted. We see 3MiB returns of
+-- {regular,elastic} tokens, and the available capacities going back to what
+-- they were. In #105185, by now we would've observed panics.
+`)
+	h.query(n1, `
+  SELECT name, crdb_internal.humanize_bytes(value::INT8)
+    FROM crdb_internal.node_metrics
+   WHERE name LIKE '%kvadmission%tokens%'
+ORDER BY name ASC;
+`)
+}
+
 type flowControlTestHelper struct {
 	t   *testing.T
 	tc  *testcluster.TestCluster

--- a/pkg/kv/kvserver/testdata/flow_control_integration/granter_admit_one_by_one
+++ b/pkg/kv/kvserver/testdata/flow_control_integration/granter_admit_one_by_one
@@ -1,0 +1,58 @@
+echo
+----
+----
+-- (Issuing regular 1024*1KiB, 3x replicated writes that are not admitted.)
+
+
+-- Flow token metrics from n1 after issuing 1024KiB, i.e. 1MiB 3x replicated writes
+-- that are yet to get admitted. We see 3*1MiB=3MiB deductions of
+-- {regular,elastic} tokens with no corresponding returns.
+SELECT name, crdb_internal.humanize_bytes(value::INT8)
+    FROM crdb_internal.node_metrics
+   WHERE name LIKE '%kvadmission%tokens%'
+ORDER BY name ASC;
+
+  kvadmission.flow_controller.elastic_tokens_available   | 21 MiB   
+  kvadmission.flow_controller.elastic_tokens_deducted    | 3.0 MiB  
+  kvadmission.flow_controller.elastic_tokens_returned    | 0 B      
+  kvadmission.flow_controller.elastic_tokens_unaccounted | 0 B      
+  kvadmission.flow_controller.regular_tokens_available   | 45 MiB   
+  kvadmission.flow_controller.regular_tokens_deducted    | 3.0 MiB  
+  kvadmission.flow_controller.regular_tokens_returned    | 0 B      
+  kvadmission.flow_controller.regular_tokens_unaccounted | 0 B      
+
+
+-- Observe the total tracked tokens per-stream on n1.
+SELECT range_id, store_id, crdb_internal.humanize_bytes(total_tracked_tokens::INT8)
+    FROM crdb_internal.kv_flow_control_handles
+
+  range_id | store_id | total_tracked_tokens  
+-----------+----------+-----------------------
+  64       | 1        | 1.0 MiB               
+  64       | 2        | 1.0 MiB               
+  64       | 3        | 1.0 MiB               
+
+
+-- (Allow below-raft admission to proceed.)
+
+
+-- Flow token metrics from n1 after work gets admitted. We see 3MiB returns of
+-- {regular,elastic} tokens, and the available capacities going back to what
+-- they were. In #105185, by now we would've observed panics.
+SELECT name, crdb_internal.humanize_bytes(value::INT8)
+    FROM crdb_internal.node_metrics
+   WHERE name LIKE '%kvadmission%tokens%'
+ORDER BY name ASC;
+
+  kvadmission.flow_controller.elastic_tokens_available   | 24 MiB   
+  kvadmission.flow_controller.elastic_tokens_deducted    | 3.0 MiB  
+  kvadmission.flow_controller.elastic_tokens_returned    | 3.0 MiB  
+  kvadmission.flow_controller.elastic_tokens_unaccounted | 0 B      
+  kvadmission.flow_controller.regular_tokens_available   | 48 MiB   
+  kvadmission.flow_controller.regular_tokens_deducted    | 3.0 MiB  
+  kvadmission.flow_controller.regular_tokens_returned    | 3.0 MiB  
+  kvadmission.flow_controller.regular_tokens_unaccounted | 0 B      
+----
+----
+
+# vim:ft=sql

--- a/pkg/util/admission/admission.go
+++ b/pkg/util/admission/admission.go
@@ -299,11 +299,22 @@ type granterWithStoreReplicatedWorkAdmitted interface {
 	// inform granters of when the write was actually done, post-admission. At
 	// admit-time we did not have sizing info for these writes, so by
 	// intercepting these writes at admit time we're able to make any
-	// outstanding token adjustments in the granter.
+	// outstanding token adjustments in the granter. When adjusting tokens, if
+	// we observe the granter was previously exhausted but is now no longer so,
+	// this interface is allowed to admit other waiting requests.
 	storeWriteDone(originalTokens int64, doneInfo StoreWorkDoneInfo) (additionalTokens int64)
 	// storeReplicatedWorkAdmittedLocked is used by below-raft admission control
 	// to inform granters of work being admitted in order for them to make any
 	// outstanding token adjustments. It's invoked with the coord.mu held.
+	// Unlike storeWriteDone, the token adjustments don't result in further
+	// admission. There are two callsites to get here:
+	// - From (*WorkQueue).granted, invoked in the GrantCoordinator loop. The
+	//   caller itself is looping, so we don't grant further admission. If we
+	//   did, our recursive callstack could be as large as the number of waiting
+	//   requests.
+	// - The fast path in (*WorkQueue).Admit, invoked by the work seeking
+	//   admission that only wants to subtract tokens. Here again there's no
+	//   need for further admission.
 	storeReplicatedWorkAdmittedLocked(originalTokens int64, admittedInfo storeReplicatedWorkAdmittedInfo) (additionalTokens int64)
 }
 

--- a/pkg/util/admission/grant_coordinator.go
+++ b/pkg/util/admission/grant_coordinator.go
@@ -156,6 +156,7 @@ func (sgc *StoreGrantCoordinators) initGrantCoordinator(storeID roachpb.StoreID)
 	coord := &GrantCoordinator{
 		settings:       sgc.settings,
 		useGrantChains: false,
+		knobs:          sgc.knobs,
 	}
 	coord.mu.numProcs = 1
 
@@ -301,7 +302,11 @@ type GrantCoordinator struct {
 	// is too low. For testing queueing behavior, we do not want the enforcement
 	// to be turned off in a non-deterministic manner so add a testing flag to
 	// disable that feature.
+	//
+	// TODO(irfansharif): Fold into the testing knobs struct below.
 	testingDisableSkipEnforcement bool
+
+	knobs *TestingKnobs
 }
 
 var _ CPULoadListener = &GrantCoordinator{}
@@ -397,9 +402,13 @@ func NewGrantCoordinators(
 	metrics := makeGrantCoordinatorMetrics()
 	registry.AddMetricStruct(metrics)
 
+	if knobs == nil {
+		knobs = &TestingKnobs{}
+	}
+
 	return GrantCoordinators{
 		Stores:  makeStoresGrantCoordinators(ambientCtx, opts, st, metrics, registry, onLogEntryAdmitted, knobs),
-		Regular: makeRegularGrantCoordinator(ambientCtx, opts, st, metrics, registry),
+		Regular: makeRegularGrantCoordinator(ambientCtx, opts, st, metrics, registry, knobs),
 		Elastic: makeElasticGrantCoordinator(ambientCtx, st, registry),
 	}
 }
@@ -438,9 +447,6 @@ func makeStoresGrantCoordinators(
 	onLogEntryAdmitted OnLogEntryAdmitted,
 	knobs *TestingKnobs,
 ) *StoreGrantCoordinators {
-	if knobs == nil {
-		knobs = &TestingKnobs{}
-	}
 	// These metrics are shared across all stores and broken down by priority for
 	// the common priorities.
 	// TODO(baptist): Add per-store metrics.
@@ -473,6 +479,7 @@ func makeRegularGrantCoordinator(
 	st *cluster.Settings,
 	metrics GrantCoordinatorMetrics,
 	registry *metric.Registry,
+	knobs *TestingKnobs,
 ) *GrantCoordinator {
 	makeRequester := makeWorkQueue
 	if opts.makeRequesterFunc != nil {
@@ -494,6 +501,7 @@ func makeRegularGrantCoordinator(
 		settings:                      st,
 		useGrantChains:                true,
 		testingDisableSkipEnforcement: opts.TestingDisableSkipEnforcement,
+		knobs:                         knobs,
 	}
 	coord.mu.grantChainID = 1
 	coord.mu.cpuOverloadIndicator = kvSlotAdjuster
@@ -893,6 +901,7 @@ func (coord *GrantCoordinator) tryGrantLocked() {
 OuterLoop:
 	for ; coord.mu.grantChainIndex < numWorkKinds; coord.mu.grantChainIndex++ {
 		localDone := false
+
 		granter := coord.granters[coord.mu.grantChainIndex]
 		if granter == nil {
 			// A GrantCoordinator can be limited to certain WorkKinds, and the

--- a/pkg/util/admission/granter.go
+++ b/pkg/util/admission/granter.go
@@ -362,16 +362,19 @@ func (cg *kvStoreTokenChildGranter) storeWriteDone(
 	cg.parent.coord.mu.Lock()
 	defer cg.parent.coord.mu.Unlock()
 	// NB: the token/metric adjustments we want to make here are the same as we
-	// want to make through the storeReplicatedWorkAdmittedLocked, so we (ab)use it.
+	// want to make through the storeReplicatedWorkAdmittedLocked, so we (ab)use
+	// it. The one difference is that post token adjustments, if we observe the
+	// granter was previously exhausted but is no longer so, we're allowed to
+	// admit other waiting requests.
 	return cg.parent.storeReplicatedWorkAdmittedLocked(
-		cg.workClass, originalTokens, storeReplicatedWorkAdmittedInfo(doneInfo))
+		cg.workClass, originalTokens, storeReplicatedWorkAdmittedInfo(doneInfo), true /* canGrantAnother */)
 }
 
 // storeReplicatedWorkAdmitted implements granterWithStoreReplicatedWorkAdmitted.
 func (cg *kvStoreTokenChildGranter) storeReplicatedWorkAdmittedLocked(
 	originalTokens int64, admittedInfo storeReplicatedWorkAdmittedInfo,
 ) (additionalTokens int64) {
-	return cg.parent.storeReplicatedWorkAdmittedLocked(cg.workClass, originalTokens, admittedInfo)
+	return cg.parent.storeReplicatedWorkAdmittedLocked(cg.workClass, originalTokens, admittedInfo, false /* canGrantAnother */)
 }
 
 func (sg *kvStoreTokenGranter) tryGet(workClass admissionpb.WorkClass, count int64) bool {
@@ -557,7 +560,10 @@ func (sg *kvStoreTokenGranter) setLinearModels(
 }
 
 func (sg *kvStoreTokenGranter) storeReplicatedWorkAdmittedLocked(
-	wc admissionpb.WorkClass, originalTokens int64, admittedInfo storeReplicatedWorkAdmittedInfo,
+	wc admissionpb.WorkClass,
+	originalTokens int64,
+	admittedInfo storeReplicatedWorkAdmittedInfo,
+	canGrantAnother bool,
 ) (additionalTokens int64) {
 	// Reminder: coord.mu protects the state in the kvStoreTokenGranter.
 	exhaustedFunc := func() bool {
@@ -576,9 +582,9 @@ func (sg *kvStoreTokenGranter) storeReplicatedWorkAdmittedLocked(
 		sg.coordMu.elasticDiskBWTokensAvailable -= additionalDiskBWTokensNeeded
 	}
 	sg.coordMu.diskBWTokensUsed[wc] += additionalDiskBWTokensNeeded
-	if additionalL0TokensNeeded < 0 || additionalDiskBWTokensNeeded < 0 {
+	if canGrantAnother && (additionalL0TokensNeeded < 0 || additionalDiskBWTokensNeeded < 0) {
 		isExhausted := exhaustedFunc()
-		if wasExhausted && !isExhausted {
+		if (wasExhausted && !isExhausted) || sg.coord.knobs.AlwaysTryGrantWhenAdmitted {
 			sg.coord.tryGrantLocked()
 		}
 	}

--- a/pkg/util/admission/granter_test.go
+++ b/pkg/util/admission/granter_test.go
@@ -143,6 +143,7 @@ func TestGranterBasic(t *testing.T) {
 				kvIOTotalTokensTaken:            metrics.KVIOTotalTokensTaken,
 				workQueueMetrics:                workQueueMetrics,
 				disableTickerForTesting:         true,
+				knobs:                           &TestingKnobs{},
 			}
 			var metricsProvider testMetricsProvider
 			metricsProvider.setMetricsForStores([]int32{1}, pebble.Metrics{})

--- a/pkg/util/admission/testing_knobs.go
+++ b/pkg/util/admission/testing_knobs.go
@@ -38,6 +38,10 @@ type TestingKnobs struct {
 	// DisableWorkQueueGranting disables the work queue from granting admission
 	// to waiting work.
 	DisableWorkQueueGranting func() bool
+
+	// AlwaysTryGrantWhenAdmitted causes the granter to unconditionally try
+	// admitting another request when admitting one.
+	AlwaysTryGrantWhenAdmitted bool
 }
 
 // ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.

